modules/fluid_properties/src/userobjects/CaloricImperfectGas.C

```diff
@@ -0,0 +1,970 @@
+  unsigned int n = std::floor(_max_temperature - _min_temperature) + 1;
+  _delta_T = (_max_temperature - _min_temperature) / ((Real)n - 1.0);
+  Real e_old;
+  Real h_old;
+  for (unsigned int j = 0; j < n; ++j)
+  {
+    Real temperature = _min_temperature + j * _delta_T;
+    Real h = _h_T->value(temperature, Point());
+    Real e = _e_T->value(temperature, Point());
+
+    if (j > 0 && (e - e_old <= 0.0 || h - h_old <= 0.0))
+      mooseError("e(T) or h(T) are not monotonically increasing with T");
+    e_old = e;
+    h_old = h;
+  }
+
+  // backward lookup tables
+  _min_e = _e_T->value(_min_temperature, Point());
+  _max_e = _e_T->value(_max_temperature, Point());
+  _delta_e = (_max_e - _min_e) / ((Real)n - 1.0);
+  _min_h = _h_T->value(_min_temperature, Point());
+  _max_h = _h_T->value(_max_temperature, Point());
+  _delta_h = (_max_h - _min_h) / ((Real)n - 1.0);
+
+  _T_h_lookup.resize(n);
+  _T_e_lookup.resize(n);
+  for (unsigned int j = 0; j < n; ++j)
+  {
+    // internal energy
+    {
+      Real internal_energy = _min_e + j * _delta_e;
+
+      // The temperature is found by e - e(T) = 0
+      Real min = _min_temperature;
+      Real max = _max_temperature;
+      auto e_diff = [&internal_energy, this](Real x)
+      { return this->e_from_p_T(0.0, x) - internal_energy; };
+      BrentsMethod::bracket(e_diff, min, max);
+      Real temperature = BrentsMethod::root(e_diff, min, max);
+      _T_e_lookup[j] = temperature;
+    }
+
+    // enthalpy
+    {
+      Real enthalpy = _min_h + j * _delta_h;
+
+      // The temperature is found by e - e(T) = 0
+      Real min = _min_temperature;
+      Real max = _max_temperature;
+      auto h_diff = [&enthalpy, this](Real x) { return this->h_from_p_T(0.0, x) - enthalpy; };
+      BrentsMethod::bracket(h_diff, min, max);
+      Real temperature = BrentsMethod::root(h_diff, min, max);
+      _T_h_lookup[j] = temperature;
+    }
+  }
+}
+
+std::string
+CaloricImperfectGas::fluidName() const
+{
+  return "caloric_imperfect_gas";
+}
+
+Real
+CaloricImperfectGas::e_from_T(Real T) const
+{
+  if (T < _min_temperature || T > _max_temperature)
+    outOfBounds();
+
+  return _e_T->value(T, Point());
+}
+
+Real
+CaloricImperfectGas::h_from_T(Real T) const
+{
+  if (T < _min_temperature || T > _max_temperature)
+    outOfBounds();
+  return _h_T->value(T, Point());
+}
+
+Real
+CaloricImperfectGas::cp_from_T(Real T) const
+{
+  if (T < _min_temperature || T > _max_temperature)
+    outOfBounds();
+  return _cp_T->value(T, Point());
+}
+
+Real
+CaloricImperfectGas::cv_from_T(Real T) const
+{
+  if (T < _min_temperature || T > _max_temperature)
+    outOfBounds();
+  return _cv_T->value(T, Point());
+}
+
+void
+CaloricImperfectGas::cv_from_T(Real T, Real & cv, Real & dcv_dT) const
+{
+  if (T < _min_temperature || T > _max_temperature)
+    outOfBounds();
+  Real pert = 1.0e-7;
+  cv = cv_from_T(T);
+  Real cv_pert = cv_from_T(T * (1 + pert));
+  dcv_dT = (cv_pert - cv) / (T * pert);
```

Comment:
This comment is moot if you delete the non-AD implementations, but if you are implementing it, you can use `_cv_T->timeDerivative()`

Reply:
`_cv_T` is now computed from `_e_T` so this won't work anymore.